crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,294 @@
+        match pairings_prod {
+            Some(mut prod) => {
+                for p in pairings.iter() {
+                    // loop through rest of list
+                    prod *= &p;
+                }
+                prod == sig.pair(&PublicKey::generator())
+            }
+            _ => pairings.is_empty(),
+        }
+    }
+}
+
+// Python Functions
+#[cfg(feature = "py-bindings")]
+#[pymethods]
+impl BLSCache {
+    #[new]
+    pub fn init() -> Self {
+        Self::generator(None)
+    }
+
+    #[staticmethod]
+    #[pyo3(name = "generator")]
+    pub fn py_generator(size: Option<&PyInt>) -> Self {
+        match size {
+            Some(s) => {
+                let usize_value: usize = s.extract::<usize>().unwrap();
+                Self::generator(Some(usize_value))
+            }
+            None => Self::generator(None),
+        }
+    }
+
+    #[pyo3(name = "aggregate_verify")]
+    pub fn py_aggregate_verify(
+        &mut self,
+        pks: &PyList,
+        msgs: &PyList,
+        sig: &Signature,
+        force_cache: &PyBool,
+    ) -> PyResult<bool> {
+        let pks_r: Vec<Bytes48> = pks
+            .iter()
+            .map(|item| item.extract::<Bytes48>())
+            .collect::<PyResult<_>>()?;
+        let msgs_r: Vec<Vec<u8>> = msgs
+            .iter()
+            .map(|item| item.extract::<Vec<u8>>())
+            .collect::<PyResult<_>>()?;
+        let force_cache_bool = force_cache.extract::<bool>()?;
+        Ok(self.aggregate_verify(&pks_r, &msgs_r, sig, force_cache_bool))
+    }
+
+    #[pyo3(name = "len")]
+    pub fn py_len(&self) -> PyResult<usize> {
+        Ok(self.cache.len())
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use crate::aggregate;
+    use crate::sign;
+    use crate::SecretKey;
+
+    #[test]
+    pub fn test_instantiation() {
+        let mut bls_cache: BLSCache = BLSCache::generator(None);
+        let byte_array: [u8; 32] = [0; 32];
+        let sk: SecretKey = SecretKey::from_seed(&byte_array);
+        let pk: PublicKey = sk.public_key();
+        let msg: [u8; 32] = [106; 32];
+        let mut aug_msg: Vec<u8> = pk.clone().to_bytes().to_vec();
+        aug_msg.extend_from_slice(&msg); // pk + msg
+        let aug_hash = hash_to_g2(&aug_msg);
+        let pairing = aug_hash.pair(&pk);
+        let mut hasher = Sha256::new();
+        hasher.update(&aug_msg);
+        let h: Bytes32 = hasher.finalize().into();
+        bls_cache.cache.put(h, pairing.clone());
+        assert_eq!(*bls_cache.cache.get(&h).unwrap(), pairing);
+    }
+
+    #[test]
+    pub fn test_aggregate_verify() {
+        let mut bls_cache: BLSCache = BLSCache::generator(None);
+        assert_eq!(bls_cache.cache.len(), 0);
+        let byte_array: [u8; 32] = [0; 32];
+        let sk: SecretKey = SecretKey::from_seed(&byte_array);
+        let pk: PublicKey = sk.public_key();
+        let msg: Vec<u8> = [106; 32].to_vec();
```

Comment:
FYI you can use `vec![106; 32]` to shortcut this. Although I'm not convinced you need all these `Vec` instances anyways. They can be potentially expensive to allocate on the heap. In tests it doesn't matter as much though.